bot/exts/events/advent_of_code/_cog.py

```diff
@@ -52,13 +60,80 @@ def __init__(self, bot: Bot):
+        Runs on a schedule, as defined in the task.loop decorator.
+        """
+        await self.bot.wait_until_guild_available()
+        guild = self.bot.get_guild(Client.guild)
+        completionist_role = guild.get_role(Roles.aoc_completionist)
+
+        aoc_name_to_member_id = {
+            aoc_name: member_id
+            for member_id, aoc_name in await self.account_links.items()
+        }
+
+        try:
+            leaderboard = await _helpers.fetch_leaderboard()
+        except _helpers.FetchingLeaderboardFailedError:
+            await self.bot.send_log("Unable to fetch AoC leaderboard during role sync.")
+            return
+
+        placement_leaderboard = json.loads(leaderboard["placement_leaderboard"])
+        members_who_deserve_role: set[discord.Member] = set()
+
+        for member_aoc_info in placement_leaderboard.values():
+            if not member_aoc_info["stars"] == 50:
+                # Only give the role to people who have completed all 50 stars
+                continue
+
+            member_id = aoc_name_to_member_id.get(member_aoc_info["name"], None)
+            if not member_id:
+                continue
+
+            member = await members.get_or_fetch_member(guild, member_id)
+            if member is None:
+                continue
+
+            if member in completionist_role.members:
+                if await self.completionist_block_list.contains(member_id):
+                    await members.handle_role_change(member, member.remove_roles, completionist_role)
+                continue
+
+            await members.handle_role_change(member, member.add_roles, completionist_role)
+            members_who_deserve_role.add(member)
+
+        for member in set(completionist_role.members) - members_who_deserve_role:
+            # Remove the role from members who have it but don't deserve it (been removed from the leaderboard)
+            await members.handle_role_change(member, member.remove_roles, completionist_role)
+
     @commands.group(name="adventofcode", aliases=("aoc",))
     @whitelist_override(channels=AOC_WHITELIST)
     async def adventofcode_group(self, ctx: commands.Context) -> None:
         """All of the Advent of Code commands."""
         if not ctx.invoked_subcommand:
             await invoke_help_command(ctx)
 
+    @with_role(Roles.admins)
+    @adventofcode_group.command(
+        name="block",
+        brief="Block a user from getting the completer role.",
+    )
+    async def block_from_role(self, ctx: commands.Context, member: discord.Member) -> None:
+        """Block the given member from receiving the AoC completionist role, removing it from them if needed."""
+        await self.bot.wait_until_guild_available()
+        completionist_role = ctx.guild.get_role(Roles.aoc_completionist)
+        if completionist_role in member.roles:
+            await member.remove_roles(completionist_role)
+
+        await self.completionist_block_list.set(member.id, None)
```

Comment:
Yea, lance doesn't have database setup, so redis is the defacto persistent storage. It's not as bad as it used to be, redis persistence that is, since it now has RDB and AOF, which we use in prod https://redis.io/topics/persistence